Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1330 @@
+integer = read <$> some numberChar
+```
+
+We now change our `atom` to also include integers.
+
+```hs
+atom :: Parser SExp
+atom = choice
+  [ SBool <$> bool
+  , SInteger <$> integer
+  ]
+```
+
+Let's also test it in GHCi:
+
+```hs
+>>> parseTest atom "42"
+SInteger 42
+
+>>> parseTest atom "hey"
+1:1:
+  |
+1 | hey
+  | ^^^
+unexpected "hey"
+expecting "false", "true", or numeric character
+```
+
+Although the error message is good, wouldn't it better if it said "integer" instead of "numeric character"? For this, we can use **labels**, which allows us to give a name to a parser. To do this, we can simply change our parser as such:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ read <$> some numberChar
+```
+
+And now the error message should be slightly improved:
+
+```hs
+>>> parseTest atom "hey"
+1:1:
+  |
+1 | hey
+  | ^^^
+unexpected "hey"
+expecting "false", "true" or integer
+```
+
+You can also use `label "boolean"` in our `bool` parser to change the message to `expecting boolean or integer`.
+
+**Exercise**: Can you change this parser so it also accepts negative integers?
+
+<details>
+  <summary>Solution</summary>
+One possible solution using `<|>`:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ read <$> (some numberChar <|> ((:) <$> char '-' <*> some numberChar))
+```
+
+However, later in the tutorial, we will see an even better way, using Megaparsec's own mechanisms.
+  <hr>
+</details>
+
+### Strings
+
+To parse a string, the idea is to find everything that is between double-quotes. This approach has a weakness, though: we won't be able to parse escaped quotes. Later in this tutorial, we will present a solution to escaped characters.
+
+To achieve this, we will use two new functions: `between` and `takeWhileP`. Using `between open close x` is equivalent to `open *> x <* close`, while `takeWhileP` will read as many characters as possible that satisfy some predicate.
+
+Note that `takeWhileP` also accepts, optionally, a label for each character that it tries to read. For now, we will leave it as `Nothing`.
+
+```hs
+str :: Parser String
+str = label "string" $ between (char '"') (char '"') (takeWhileP Nothing (/= '"'))
+-- Or:
+--- str = label "string" $ char '"' *> takeWhileP Nothing (/= '"') <* char '"'
+```
+
+In `atom`, include `SString <$> str`.
+
+```hs
+>>> parseTest atom "\"hey\""
+SString "hey"
+```
+
+### Identifiers
+
+An identifier will be like a variable in a programming language. We can choose the naming convention for such a thing, but here I chose the ones that are frequently used for C-like languages: the first letter must be a letter or an underscore, while the remainder may be a letter, digit, or underscore.
+
+Here we use the `many` function, which is very similar to the `some` function discussed before. The difference is that `any` tries to run the given parser **0 or more times**. In Regex, our `identifier` parser is equivalent to `[a-zA-Z][a-zA-Z0-9]*`.
```

Comment:
Here and elsewhere, I don't think there is a need to capitalize regex.

Reply:
This was a bit surprising to me, but it makes sense, it's just "regular expressions" which shouldn't be capitalized... I've changed it.